### PR TITLE
Add flag --recompile

### PR DESCRIPTION
**Pull request by bartlomieju**
First attempt to solve #790. It's quite hacky but it does the job and forces compilation.

Don't know yet how to test it. Feedback welcome.

EDIT: typo

### CODE REVIEW

File: src/flags.rs [hunk 1]
```diff
@@ -32,6 +33,7 @@ pub fn print_usage() {
 --allow-env        Allow environment access.
 -v or --version    Print the version.
 -r or --reload     Reload cached remote resources.
+-c or --compile    Force recompilation of TypeScript code.
```
Comment (hayd) [inline src/flags.rs]: Taking the -c flag seems unnecessary, is this something users are going to be doing often enough to merit this abbreviation?

Comment (ry) [reply src/flags.rs]: It’s a good point. We should try to be conservative with the number of flags added. Maybe —recompile is better?

Comment (bartlomieju) [reply src/flags.rs]: I agree, I've used `-c` and `--compile` as per issue.

**Comment by bartlomieju** [comment]
I have renamed the flag, but TBH I am not satisfied how the flag is passed down.

Maybe we should add `configure` or `setFlags` methods to compiler that would be called before `run` to save flags on the compiler instance instead of passing *optional* argument down 3 functions?

File: js/compiler.ts [hunk 2]
```diff
@@ -196,7 +196,10 @@ export class DenoCompiler
    * Get the dependencies for a given module, but don't run the module,
    * just add the module factory to the run queue.
    */
-  private _gatherDependencies(moduleMetaData: ModuleMetaData): void {
+  private _gatherDependencies(
+      moduleMetaData: ModuleMetaData,
+      forceCompilation?: boolean
```
Comment (ry) [inline js/compiler.ts]: `s/forceCompilation/recompile/` ? Just to stick with the same nomenclature?

Comment (bartlomieju) [reply js/compiler.ts]: I got rid of that argument entirely.

File: js/compiler.ts [hunk 3]
```diff
@@ -55,6 +55,10 @@ type OutputCode = string;
  * The original source code
  */
 type SourceCode = string;
+/**
+ * Flags that are passed from CLI arguments
+ */
+type CompilerFlags = { [key: string]: boolean };
```
Comment (bartlomieju) [inline js/compiler.ts]: @ry I guess we will have to replace `boolean` with `any` pretty soon, but it should be enough for now.

Comment (ry) [reply js/compiler.ts]: @bartlomieju I think these are TS compiler flags - so recompile shouldn’t go in there... but maybe @kitsonk can suggest the best place to put this flag.

Comment (ry) [reply js/compiler.ts]: Sorry I misunderstood this when reviewing it the first time. I see that you're adding this now.

I'd prefer if you did not generalize - I think a simple `recompile: bool` member would work. When there are more than 3 compiler flags we can discuss generalizing.

File: js/compiler_test.ts [hunk 4]
```diff
@@ -458,6 +458,25 @@ test(function compilerGetScriptFileNames() {
...
+      recompile: true,
+  });
+  compilerInstance.run("foo/bar.ts", "/root/project");
+  assertEqual(getEmitOutputStack.length, 2, "Expected two emitted file.");
+  assert(
+      getEmitOutputStack[0] === getEmitOutputStack[1],
+      "Expected same file to be emitted twice."
+  );
+  teardown();
+});
```
Comment (ry) [inline js/compiler_test.ts]: Cool - thanks!

File: js/compiler.ts [hunk 5]
```diff
@@ -169,6 +173,8 @@ export class DenoCompiler
   // A reference to the global scope so it can be monkey patched during
   // testing
   private _window = window;
+  // Flags provide settings that can be changed via CLI arguments
+  private _flags: CompilerFlags = {};
```
Comment (ry) [inline js/compiler.ts]: Replace with 
```ts
  recompile: bool;
```

File: js/compiler.ts [hunk 6]
```diff
@@ -669,6 +679,10 @@ export class DenoCompiler
     });
   }
 
+  setFlags(flags: CompilerFlags): void {
+    this._flags = { ...this._flags, ...flags };
+  }
```
Comment (ry) [inline js/compiler.ts]: Remove.

**Comment by bartlomieju** [comment]
@ry I've used your hints. Thanks for the review. Let me know if I should change anything else.

**Pull request merged**